src/routes/safe/components/Apps/ManageApps.js

```diff
@@ -0,0 +1,183 @@
+  margin: 0;
+`
+const APP_INFO = { iconUrl: appsIconSvg, name: '', error: false }
+
+type Props = {
+  appList: Array<{
+    id: string,
+    iconUrl: string,
+    name: string,
+    disabled: boolean,
+  }>,
+  onAppAdded: (app: any) => void,
+  onAppToggle: (appId: string, enabled: boolean) => void,
+}
+
+const ManageApps = ({ appList, onAppAdded, onAppToggle }: Props) => {
+  const [isOpen, setIsOpen] = useState(false)
+
+  const [appInfo, setAppInfo] = useState(APP_INFO)
+  const [isSubmitDisabled, setIsSubmitDisabled] = useState(true)
+
+  const onItemToggle = (itemId: string, checked: boolean) => {
+    onAppToggle(itemId, checked)
+  }
+
+  const handleSubmit = () => {
+    setIsOpen(false)
+    onAppAdded(appInfo)
+  }
+
+  const cleanAppInfo = () => setAppInfo(APP_INFO)
+
+  const urlValidator = (value: string) => {
+    return /(?:^|[ \t])((https?:\/\/)?(?:localhost|[\w-]+(?:\.[\w-]+)+)(:\d+)?(\/\S*)?)/gm.test(value)
+      ? undefined
+      : 'Please, provide a valid url'
+  }
+
+  const safeAppValidator = async (value) => {
+    const appInfo = await getAppInfoFromUrl(value)
+
+    if (appInfo.error) {
+      setAppInfo(APP_INFO)
+      return 'This is not a valid Safe app.'
+    }
+
+    setAppInfo({ ...appInfo })
+  }
+
+  const uniqueAppValidator = (value) => {
+    const exists = appList.find((a) => a.url === value.trim())
+    return exists ? 'This app is already registered.' : undefined
+  }
+
+  const onFormStatusChange = ({ pristine, valid, validating }) => {
+    if (!pristine) {
+      setIsSubmitDisabled(validating || !valid || appInfo.error)
+    }
+  }
+
+  const getAddAppForm = () => {
+    return (
+      <GnoForm
+        initialValues={{
+          appUrl: '',
+          agreed: false,
+        }}
+        onSubmit={handleSubmit}
+        testId={FORM_ID}
+      >
+        {() => (
+          <>
+            <StyledText size="xl">Add custom app</StyledText>
+            <DebounceValidationField
+              component={TextField}
+              label="App URL"
+              name="appUrl"
+              placeholder="App URL"
+              type="text"
+              validate={composeValidators(cleanAppInfo, required, urlValidator, uniqueAppValidator, safeAppValidator)}
+            />
+
+            <AppInfo>
+              <Img alt="Token image" height={55} src={appInfo.iconUrl} />
+              <StyledTextFileAppName label="App name" readOnly value={appInfo.name} />
+            </AppInfo>
+
+            <FormSpy
+              onChange={onFormStatusChange}
+              subscription={{
+                valid: true,
+                pristine: true,
+                validating: true,
+              }}
+            />
+
+            <Field
+              component={StyledCheckbox}
+              label="I agree to use this app at my own risk"
+              name="agreed"
+              type="checkbox"
+              validate={(value) => (value === true ? undefined : 'Required')}
```

Comment:
we have required validator in `src/components/forms/validator.js`, could it be used here?

Reply:
done